src/read_write/node_iterator.rs

```diff
@@ -17,25 +17,39 @@ use std::io::Result;
 
 pub trait NodeReader {
     fn read(&mut self) -> Result<Point>;
-    fn num_points(&self) -> usize;
 }
 
 /// Streams points from our data provider representation.
-pub enum NodeIterator<R> {
-    WithData(R),
-    Empty,
+pub struct NodeIterator<R> {
+    reader: Option<R>,
+    num_points: usize,
+    point_count: usize,
+}
+
+impl<R> Default for NodeIterator<R> {
+    fn default() -> Self {
+        NodeIterator {
+            reader: None,
+            num_points: 0,
+            point_count: 0,
+        }
+    }
 }
 
 impl<R> NodeIterator<R>
 where
     R: NodeReader,
 {
-    pub fn new(reader: R) -> Self {
-        if reader.num_points() == 0 {
-            return NodeIterator::Empty;
+    pub fn new(reader: R, num_points: usize) -> Self {
```

Comment:
Could you comment on why this changed? Does that mean you have to declare beforehand how many points you want to read?

Reply:
Unfortunately this has always been there, see e.g. https://github.com/googlecartographer/point_cloud_viewer/blob/master/src/octree/generation.rs#L195 for its usage. I actually just moved it from the reader to the iterator, so at least the reader doesn't need to know it beforehand...

Reply:
Ah, okay.